language/bytecode-verifier/src/borrow_graph.rs

```diff
@@ -0,0 +1,272 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module defines the (acyclic) borrow graph for the type and memory safety analysis.
+//! A node in the borrow graph represents an abstract reference.  Each edge in the borrow graph
+//! is labeled with a (possibly empty) sequence of label elements.  A label element is either a
+//! field index or a local index or a struct index.  An edge coming out of frame_root
+//! (see abstract_state.rs) is labeled with a sequence beginning with a local or struct index
+//! followed by zero or more field indices.  An edge coming out of a node different from frame_root
+//! is labeled by a sequence of zero or more field indices.
+
+//! An edge in the borrow graph from a node other than frame_root represents a prefix relationship
+//! between the references represented by the source and sink of the edge.  There are two kinds of
+//! edges---strong and weak.  A strong edge from node a to b labeled by sequence p indicates that
+//! b is equal to the p-extension of a.  Instead, if the edge was weak, it indicates that b is an
+//! extension of the p-extension of a.
+
+use crate::nonce::Nonce;
+use mirai_annotations::{
+    checked_assume, checked_postcondition, checked_precondition, checked_verify,
+};
+use std::collections::{BTreeMap, BTreeSet};
+
+/// The type of an edge
+#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
+enum EdgeType {
+    Strong,
+    Weak,
+}
+
+/// The label on an edge
+pub type Label<T> = Vec<T>;
+
+/// A labeled edge in borrow graph
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Edge<T> {
+    edge_type: EdgeType,
+    label: Label<T>,
+    to: Nonce,
+}
+
+impl<T> Edge<T>
```

Comment:
Maybe:
```suggestion
impl<T:Ord> Ord for Edge<T> {
```
... and `compare()` on line 254?

Reply:
I looked at the documentation for the trait Ord.  It says the Ord is a total order.  The prefix relationship that I want is a partial order; it is possible for two edges not to be related.